prover/src/constraints/composition_poly.rs

```diff
@@ -90,22 +101,27 @@ impl<E: FieldElement> CompositionPoly<E> {
 
 /// Splits polynomial coefficients into the specified number of columns. The coefficients are split
 /// in such a way that each resulting column has the same degree. For example, a polynomial
-/// a * x^3 + b * x^2 + c * x + d, can be rewritten as: (b * x^2 + d) + x * (a * x^2 + c), and then
-/// the two columns will be: (b * x^2 + d) and (a * x^2 + c).
-fn transpose<E: FieldElement>(coefficients: Vec<E>, num_columns: usize) -> Vec<Vec<E>> {
-    let column_len = coefficients.len() / num_columns;
-
-    let mut result = unsafe {
-        (0..num_columns)
-            .map(|_| uninit_vector(column_len))
-            .collect::<Vec<_>>()
+/// a * x^3 + b * x^2 + c * x + d, can be rewritten as: (c * x + d) + x^2 * (a * x + b), and then
+/// the two columns will be: (c * x + d) and (a * x + b).
+fn segment<E: FieldElement>(coefficients: Vec<E>, trace_len: usize) -> Vec<Vec<E>> {
+    let degree_plus_1 = degree_of(&coefficients) + 1;
+    let remainder = degree_plus_1 % trace_len;
+
+    let num_columns = if remainder == 0 {
+        degree_plus_1 / trace_len
+    } else {
+        (degree_plus_1 / trace_len) + 1
     };
```

Comment:
Is there an easy way to compute the number of columns without using `degree_of()`? This function will scan memory from the end of the vector until a non-zero element is found. Not sure if it'll have significant impact on performance, but if there is an easy way to avoid it (i.e., buy reading number of columns from AIR context), we should do it).

Reply:
~The problem with using number of columns is that `degree_plus_1` doesn't divide evenly by `trace_len` and hence we cant replace `for (i, coeff) in coefficients.iter().take(degree_plus_1).enumerate()` with `for (i, coeff) in coefficients.iter().take(num_cols * trace_len).enumerate()`.~
Actually, it will work, provided we initialize the result matrix with `E::ZERO`.
I have made an attempt at using the AIR context, let me know if it does the job.

Reply:
Actually it will work even without intialization using your suggestion with `split_off`